Zend/zend_attributes.c

```diff
@@ -70,6 +114,74 @@ ZEND_API zend_attribute *zend_get_parameter_attribute_str(HashTable *attributes,
 	return get_attribute_str(attributes, str, len, offset + 1);
 }
 
+ZEND_API int zend_get_attribute_value(zval *ret, zend_attribute *attr, uint32_t i, zend_class_entry *scope)
+{
+	if (i >= attr->argc) {
+		return FAILURE;
+	}
+
+	ZVAL_COPY_OR_DUP(ret, &attr->argv[i]);
+
+	if (Z_TYPE_P(ret) == IS_CONSTANT_AST) {
+		if (SUCCESS != zval_update_constant_ex(ret, scope)) {
+			zval_ptr_dtor(ret);
+			return FAILURE;
+		}
+	}
+
+	return SUCCESS;
+}
+
+static const char *target_names[] = {
+	"CLASS",
+	"FUNCTION",
+	"METHOD",
+	"PROPERTY",
+	"CLASS CONSTANT",
+	"PARAMETER"
+};
+
+ZEND_API int zend_get_attribute_target_names(char *buf, uint32_t flags)
```

Comment:
I'd prefer accepting a `smart_str` here and appending to it.

Reply:
I changed the function to return a `zend_string` instead. It used a `smart_str` internally to assemble the string.